### PR TITLE
Add ItemStack aware version of onBlockPlaced

### PR DESCRIPTION
**Pull request by shadowfacts**
Allows for setting the block state based on non-metadata `ItemStack` such as NBT or capabilities.

All Vanilla calls to the original `onBlockPlaced` now call the `ItemStack` aware variant which by default calls the (now deprecated) original one.

### CODE REVIEW

**Comment by hea3ven** [comment]
Can't you do it in onBlockPlacedBy?

**Comment by williewillus** [comment]
onBlockPlacedBy runs *after* the state is set

**Comment by williewillus** [comment]
onBlockPlaced is actually a terrible name, what it actually does is "get blockstate from stack"

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 1]
```diff
@@ -143,7 +143,21 @@
      }
  
      public boolean func_180639_a(World p_180639_1_, BlockPos p_180639_2_, IBlockState p_180639_3_, EntityPlayer p_180639_4_, EnumHand p_180639_5_, @Nullable ItemStack p_180639_6_, EnumFacing p_180639_7_, float p_180639_8_, float p_180639_9_, float p_180639_10_)
-@@ -674,25 +681,35 @@
+@@ -633,6 +640,13 @@
+     {
+     }
+ 
++    public IBlockState onBlockPlaced(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, ItemStack stack, EntityLivingBase p_180642_8_)
```
Comment (kashike) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: Should really be adding `ItemStack` as the *last* parameter.

**Comment by shadowfacts** [comment]
1. @hea3ven As willie said, `onBlockPlacedBy` runs after the block is placed whereas `onBlockPlaced` just gets the correct blockstate from the stack. I could call `World#setBlockState` in `onBlockPlacedBy` to set the correct state but that would just be redundant.
2. The `ItemStack` parameter has been moved to the end of the parameter list

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 2]
```diff
@@ -143,7 +143,21 @@
      }
  
      public boolean func_180639_a(World p_180639_1_, BlockPos p_180639_2_, IBlockState p_180639_3_, EntityPlayer p_180639_4_, EnumHand p_180639_5_, @Nullable ItemStack p_180639_6_, EnumFacing p_180639_7_, float p_180639_8_, float p_180639_9_, float p_180639_10_)
-@@ -674,25 +681,35 @@
+@@ -633,6 +640,13 @@
+     {
+     }
+ 
++    public IBlockState onBlockPlaced(World worldIn, BlockPos pos, EnumFacing facing, float hitX, float hitY, float hitZ, int meta, EntityLivingBase placer, ItemStack stack)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: Umm no, do this properly.

Comment (shadowfacts) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: What do you mean by "properly"?

**Comment by shadowfacts** [comment]
Switched from overloading `onBlockPlaced` to a new method, `getStateForPlacement`.

**Comment by cpw** [comment]
This pull request should be updated and refreshed against the 1.10.x branch.

**Pull request closed without merging**